thingsboard_gateway/gateway/tb_gateway_service.py

```diff
@@ -700,44 +700,48 @@ def __send_to_storage(self):
                                             "deviceType": data['deviceType'],
                                             "attributes": {},
                                             "telemetry": []}
+                            empty_adopted_data_size = self.__get_data_size(adopted_data)
+                            adopted_data_size = empty_adopted_data_size
 
                             # First, loop through the attributes
                             for attribute in data['attributes']:
                                 adopted_data['attributes'].update(attribute)
-                                adopted_data_size = self.__get_data_size(adopted_data)
+                                adopted_data_size += self.__get_data_size(attribute)
                                 if adopted_data_size >= max_data_size:
                                     # We have surpassed the max_data_size, so send what we have and clear attributes
                                     self.__send_data_pack_to_storage(adopted_data, connector_name)
                                     adopted_data['attributes'] = {}
-                            
+                                    adopted_data_size = empty_adopted_data_size
+
                             # Now, loop through telemetry. Possibly have some unsent attributes that have been adopted.
                             telemetry = data['telemetry'] if isinstance(data['telemetry'], list) else [data['telemetry']]
+                            ts_to_index = {}
                             for ts_kv_list in telemetry:
                                 ts = ts_kv_list['ts']
                                 for kv in ts_kv_list['values']:
-                                    if len(adopted_data['telemetry']) == 0:
-                                        adopted_data['telemetry'] = [
-                                            {'ts': ts, 'values': {kv: ts_kv_list['values'][kv]}}]
+                                    if ts in ts_to_index:
+                                        kv_data = {kv: ts_kv_list['values'][kv]}
+                                        adopted_data['telemetry'][ts_to_index]['values'].update(kv_data)
```

Comment:
Raises exception: "list indices must be integers or slices, not dict"
has to be fixed to: `adopted_data['telemetry'][ts_to_index[ts]]['values'].update(kv_data)`

Reply:
Fixed one issue, create another( Thanks and fixed